### PR TITLE
Use GN & upgrade to Chromium 54

### PR DESCRIPTION
**Pull request by alespergl**
This is an extension of #249 which in addition to migrating to GN for building also upgrades to Chromium 54.

I basically just cherry picked all work that's been done in the `chrome54` branch which was related to the Chromium upgrade and applied it on top of the original `use_gn` branch.

### CODE REVIEW

**Comment by zcbenz** [comment]
I just changed the branch to download source tarball instead of using depot tools to sync source code, because of some issues with our CI servers (the Mac machine is running out of disk space, and the `fetch` script just hangs on the Windows machine).

Will revert it when we fixed the CI servers in future.

**Comment by zcbenz** [comment]
@alespergl I don't know why it failed, maybe it is related to VS and Windows SDK's versions, but only patching `bluetooth_task_manager_win.cc` would get [compilation error](http://208.52.191.140:8080/job/libchromiumcontent-win/184/console) when building for ia32 on CI machine.

File: chromiumcontent/BUILD.gn [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+             "-s", rebase_path(outputs[0]) ]
+
+    deps = [ ":targets" ]
+  }
+
+  # Normally we would just put everything in a single library plus V8. But
+  # toolchains have various limitations that we would hit - e.g. MSVC cannot
+  # produce a .lib file bigger than 4 GB, LLVM has a limit on the number of
+  # object files on the command line. So we produce multiple libraries.
+  group("libs") {
```
Comment (bridiver) [inline chromiumcontent/BUILD.gn]:
```
diff --git a/build/toolchain/mac/BUILD.gn b/build/toolchain/mac/BUILD.gn
index 0c00e995fffe139ac39badd12790acbad3a78a7a..e5651111ce284f04ce3b5d6157be84c3738e732c 100644
--- a/build/toolchain/mac/BUILD.gn
+++ b/build/toolchain/mac/BUILD.gn
@@ -210,13 +210,15 @@ template("mac_toolchain") {
     }
 
     tool("alink") {
+      rspfile = "{{output}}.rsp"
       script =
           rebase_path("//build/toolchain/mac/filter_libtool.py", root_build_dir)
-      command = "rm -f {{output}} && TOOL_VERSION=${tool_versions.filter_libtool} python $script libtool -static {{arflags}} -o {{output}} {{inputs}}"
+      command = "rm -f {{output}} && TOOL_VERSION=${tool_versions.filter_libtool} python $script libtool -static {{arflags}} -o {{output}} -filelist \"$rspfile\""
       description = "LIBTOOL-STATIC {{output}}"
       outputs = [
         "{{output_dir}}/{{target_output_name}}{{output_extension}}",
       ]
+      rspfile_content = "{{inputs_newline}}"
       default_output_dir = "{{target_out_dir}}"
       default_output_extension = ".a"
       output_prefix = "lib"
```

Comment (bridiver) [inline chromiumcontent/BUILD.gn]: that fixes the .o file limit on the command line. I'm going to submit a patch to chromium for it

Comment (bridiver) [inline chromiumcontent/BUILD.gn]: you might want to look at `split_static_library` for windows

Comment (alespergl) [inline chromiumcontent/BUILD.gn]: Interesting. 👍 If your patch gets merged to Chromium, we can surely combine some of the libs. Right now I don't see a benefit which would warrant adding this patch to libcc. And the MSVC limitation remains.

**Comment by kevinsawicki** [comment]
Rebased this and pulled in a few new patches from master and updated them to 54.

**Comment by kevinsawicki** [comment]
1.4.15 is now out so I'm going to merge this and get it landed on Electron master.

Thanks for all your awesome work on this @alespergl 👍 👍 👍

**Pull request merged**